src/BaseGen.sol

```diff
@@ -153,13 +158,21 @@ contract BaseGen is ERC721, ERC721Burnable, ERC721Royalty, Ownable {
     }
 
     function _splitPayment(uint256 amount) private {
+        _splitPayment(amount, address(0));
+    }
+
+    function _splitPayment(uint256 amount, address referrer) private {
+        uint256 receiverAmount = amount / 2;
         if (_receiver != address(0)) {
-            uint256 splitAmount = amount / 2;
-            payable(_receiver).transfer(splitAmount);
-            payable(owner()).transfer(splitAmount);
-        } else {
-            payable(owner()).transfer(amount);
+            payable(_receiver).transfer(receiverAmount);
+        }
+        if (referrer != address(0) && referrer != _receiver) {
+            (, uint256 referrerAmount) = royaltyInfo(0, amount);
+            if (referrerAmount > 0 && referrerAmount < address(this).balance) {
+                payable(referrer).transfer(referrerAmount);
+            }
         }
+        payable(owner()).transfer(address(this).balance);
```

Comment:
it doesn't feel right to use balance as what if minter will send more than needed?
Maybe we should make it equal here?  L147

if it's okay, than probably it's a best way